src/cpu/flexcpu/simple_dataflow_cpu.cc

```diff
@@ -0,0 +1,160 @@
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Authors: Bradley Wang
+ */
+
+#include "cpu/flexcpu/simple_dataflow_cpu.hh"
+
+using namespace std;
+
+SimpleDataflowCPU::SimpleDataflowCPU(SimpleDataflowCPUParams* params):
+    BaseCPU(params),
+    _dataPort(name() + "._dataPort", this),
+    _instPort(name() + "._instPort", this)
+{
+    fatal_if(FullSystem, "FullSystem not implemented for SimpleDataflowCPU");
+
+    for (unsigned i = 0; i < numThreads; i++) {
+        // Move constructor + vector should allow us to avoid memory leaks like
+        // the SimpleCPU implementations have.
+        threads.push_back(make_unique<SDCPUThread>(this, i, params->system,
+            params->workload[i], params->itb, params->dtb, params->isa[i]));
+        threadContexts.push_back(threads[i].getThreadContext());
+    }
+    // TODO add any other constructor details
+}
+
+SimpleDataflowCPU::~SimpleDataflowCPU()
+{
+    //
+}
+
+void
+SimpleDataflowCPU::activateContext(ThreadID tid)
+{
+    //DPRINTF(SimpleDataflowCPU, "ActivateContext %d\n", tid);
+
+    BaseCPU::activateContext(tid);
+    // TODO activate the thread
+
+    // TODO schedule the first fetch event.
+    // schedule(new EventFunctionWrapper([this]{ fetch(); },
+    //                                   name() + ".initial_fetch",
+    //                                   true),
+    //          curTick());
+}
+
+MasterPort&
+SimpleDataflowCPU::getDataPort() {
+    return _dataPort;
+}
+
+MasterPort&
+SimpleDataflowCPU::getInstPort() {
+    return _instPort;
+}
+
+void
+SimpleDataflowCPU::init()
+{
+    BaseCPU::init();
+
+    if (!params()->switched_out &&
+        system->getMemoryMode() != Enums::timing)
+    {
+        fatal("The Dataflow CPU requires the memory system to be in "
+            "'timing' mode.\n");
+    }
+
+    for (SDCPUThread& thread : threads) {
+        ThreadContext* tc = thread.getThreadContext();
+        tc->initMemProxies(tc);
+
+        // TODO if fullsystem, we need to check if we need to call initCPU on
+        // the ISA
+    }
+    // TODO add post-construction initialization code
+}
+
+void
+SimpleDataflowCPU::startup()
+{
+    BaseCPU::startup();
+
+    for (SDCPUThread& thread : threads) {
+        thread.startup();
+    }
+    // TODO add code which needs to be run during the first simulation event
+}
+
+void
+SimpleDataflowCPU::switchOut()
+{
+    // TODO flush speculative state.
+    // TODO clear any outstanding memory requests.
```

Comment:
This is called drain, BTW, and there's a special API for it. In this function, you should just assert that this object *is* drained.